src/main/java/hudson/plugins/sshslaves/SSHLauncher.java

```diff
@@ -1338,7 +1356,22 @@ private boolean isRecoverable(String message) {
      * {@inheritDoc}
      */
     @Override
-    public synchronized void afterDisconnect(SlaveComputer slaveComputer, final TaskListener listener) {
+    public void afterDisconnect(SlaveComputer slaveComputer, final TaskListener listener) {
+        if (connection == null) {
+            // Nothing to do here, the connection is not established
+            return;
+        }
+
+        ExecutorService srv = launcherExecutorService;
+        if (srv != null) {
+            // If the service is still running, shut it down and interrupt the operations if any
+            srv.shutdown();
+        }
+
+        tearDownConnection(slaveComputer, listener);
+    }
+
+    private synchronized void tearDownConnection(@Nonnull SlaveComputer slaveComputer, final @Nonnull TaskListener listener) {
```

Comment:
Why are you introducing this method that is called only once? Could remove ~4 diff lines.

Reply:
Otherwise I would have to use `synchronized(this) {` which would tabulate the code. And that's what you asked to avoid in https://github.com/jenkinsci/ssh-slaves-plugin/pull/75#discussion_r152390592